src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -240,149 +240,187 @@ private static ImmutableArray<TextChangeRange> Merge(IReadOnlyList<ImmutableArra
+
+            while (oldIndex < oldChanges.Length)
             {
-                // no more new changes, just add remaining old changes
-                while (oldIndex < oldChanges.Length)
-                {
-                    AddRange(list, oldChanges[oldIndex]);
-                    oldIndex++;
-                }
+                add(oldChange); // no need to adjust oldDelta at this point
+                tryGetNext(oldChanges, ref oldIndex, out oldChange);
             }
 
-            return list.ToImmutableArray();
-        }
+            while (newIndex < newChanges.Length)
+            {
+                addAdjustedNewChange(newChange);
+                tryGetNext(newChanges, ref newIndex, out newChange);
+            }
 
-        private static void AddRange(List<TextChangeRange> list, TextChangeRange range)
-        {
-            if (list.Count > 0)
+            return builder.ToImmutableAndFree();
+
+            static bool tryGetNext(ImmutableArray<TextChangeRange> changes, ref int index, out TextChangeRange next)
             {
-                var last = list[list.Count - 1];
-                if (last.Span.End == range.Span.Start)
+                if (++index < changes.Length)
                 {
-                    // merge changes together if they are adjacent
-                    list[list.Count - 1] = new TextChangeRange(new TextSpan(last.Span.Start, last.Span.Length + range.Span.Length), last.NewLength + range.NewLength);
-                    return;
+                    next = changes[index];
+                    return true;
                 }
                 else
                 {
-                    Debug.Assert(range.Span.Start > last.Span.End);
+                    next = default;
+                    return false;
                 }
             }
 
-            list.Add(range);
+            void addAndAdjustOldDelta(TextChangeRange change)
+            {
+                oldDelta = oldDelta - change.Span.Length + change.NewLength;
+                add(change);
+            }
+
+            void addAdjustedNewChange(TextChangeRange change)
+            {
+                add(new TextChangeRange(new TextSpan(change.Span.Start - oldDelta, change.Span.Length), change.NewLength));
+            }
+
+            void add(TextChangeRange range)
+            {
+                if (builder.Count > 0)
+                {
+                    var last = builder[^1];
+                    if (last.Span.End == range.Span.Start)
+                    {
+                        // merge changes together if they are adjacent
+                        builder[^1] = new TextChangeRange(new TextSpan(last.Span.Start, last.Span.Length + range.Span.Length), last.NewLength + range.NewLength);
+                        return;
+                    }
+                    else
+                    {
+                        // changes must not overlap
+                        Debug.Assert(range.Span.Start > last.Span.End);
```

Comment:
Contract vs assert.  we want to know immediately if we violate this.

Reply:
Done